### PR TITLE
Add understand chapter for HIP Graphs

### PR DESCRIPTION
**Pull request by MKKnorr**

### CODE REVIEW

**Comment by neon60** [comment]
@MKKnorr CI fails, please fix it.

**Comment by MKKnorr** [comment]
> @MKKnorr CI fails, please fix it.

That was just a timeout due to some downloads in the linting failing. The checks passed before. Triggering the pipeline again fixed it

File: docs/understand/hipgraph.rst [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+.. note::
+    The HIP graph API is currently in Beta. Some features can change and might
+    have outstanding issues. Not all features supported by CUDA graphs are yet
+    supported. For a list of all currently supported functions see the
+    :doc:`HIP graph API documentation<../doxygen/html/group___graph>`.
+
+Graph format
+================================================================================
+
+A HIP graph is, like any other graph, made up of nodes and edges. The nodes of a
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: A HIP graph is made up of nodes and edges. The nodes of a

File: docs/understand/hipgraph.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+================================================================================
+
+The standard way of launching work on GPUs via streams incurs a small overhead for each operation involved
+every time. For kernels that take a considerable amount to finish, this overhead
+usually is negligible, however many workloads, including scientific simulations
+and AI, involve launching many relatively small kernels repeatedly for many iterations.
+
+HIP graphs have been specifically designed to tackle this problem by only
+requiring one launch from the host per iteration, and minimizing that overhead
+by performing most of the initialization beforehand. Graphs may provide
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: by performing most of the initialization beforehand. Graphs can provide

File: docs/understand/hipgraph.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+.. figure:: ../data/understand/hipgraph/hip_graph.svg
+    :alt: Diagram depicting the difference between using streams to execute
+          kernels with dependencies, resolved by explicitly calling
+          hipDeviceSynchronize, or using graphs, where the edges denote the
+          dependencies.
+
+HIP graph advantages
+================================================================================
+
+The standard way of launching work on GPUs via streams incurs a small overhead for each operation involved
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: The standard way of launching work on GPUs via streams incurs a small overhead for each iteration of the operation involved.

File: docs/understand/hipgraph.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+every time. For kernels that take a considerable amount to finish, this overhead
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: For kernels that perform large operations during an iteration, this overhead

File: docs/understand/hipgraph.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+usually is negligible, however many workloads, including scientific simulations
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: is usually negligible. However, for many workloads such as scientific simulations

File: docs/understand/hipgraph.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+and AI, involve launching many relatively small kernels repeatedly for many iterations.
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: and AI, a kernel performs a small operation for many iterations, and so the overhead of launching kernels can be a significant cost on performance.

File: docs/understand/hipgraph.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+As HIP graphs require some set up and initialization overhead before their first
+execution, they only provide a benefit for workloads that require many iterations to complete.
+
+Setting up HIP graphs
+================================================================================
+
+HIP graphs can be created by explicitly defining them, or using stream capture.
+For the available functions see the
+:doc:`HIP graph API documentation<../doxygen/html/group___graph>`.
```
Comment (randyh62) [inline docs/understand/hipgraph.rst:84]: The ending seems a little sudden.  I think an example graph would be great to have here.

**Review by randyh62 — APPROVED** [review]
In general I think an example would be a great addition, even if it was just pseudocode.

File: docs/understand/hipgraph.rst (hunk 7 above)
Comment (MKKnorr) [reply docs/understand/hipgraph.rst:84]: What about the figure earlier in this document comparing launching kernels via streams and the graph next to it? I think it fits better up there.

Also the goal would be to have a how-to page, that would be linked to here. However that section is not finished, as I don't have the means to test the code currently. Also we needed to reconsider how to handle this, as the graph API is still in beta.

**Comment by MKKnorr** [comment]
> In general I think an example would be a great addition, even if it was just pseudocode.

I agree, but that's a better fit for a how-to page. As stated in another comment, that section will most likely not be ready soon. A tutorial is also planned, on top of this documentation

File: docs/data/understand/hipgraph/hip_graph.svg
Comment (neon60) [inline docs/data/understand/hipgraph/hip_graph.svg:1]: I do not see the benefit of hipGraph over explicitly calling hipDeviceSynchronize. Please update the image.

Comment (neon60) [reply docs/data/understand/hipgraph/hip_graph.svg:1]: Now I see.

**Review by neon60 — APPROVED** [review]

File: docs/data/understand/hipgraph/hip_graph.svg
Comment (randyh62) [inline docs/data/understand/hipgraph/hip_graph.svg:1]: This image could use a little green, or blue, or yellow.

File: docs/how-to/programming_manual.md [hunk 8]
```diff
@@ -146,7 +146,7 @@ For Linux developers, the link [here](https://github.com/ROCm/hip-tests/blob/dev
 
 ## HIP Graph
```
Comment (randyh62) [inline docs/how-to/programming_manual.md:147]: Is it necessary to have this here? it is not much more than a reference, which shows up two or three times in the index or TOC.

File: docs/understand/hipgraph.rst [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    :description: This chapter provides an overview over the usage of HIP graph.
+    :keywords: ROCm, HIP, graph, stream
+
+.. understand_HIP_graph:
+
+********************************************************************************
+HIP graph
+********************************************************************************
+
+HIP graphs are an alternative way of executing work on a GPU. It can provide
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: I would eliminate this first paragraph, leave the following Beta note in place, and start this _Understanding HIP Graphs_ topic with the _Graph format_ text (but without the header).

File: docs/understand/hipgraph.rst [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+          kernels with dependencies, resolved by explicitly calling
+          hipDeviceSynchronize, or using graphs, where the edges denote the
+          dependencies.
+
+HIP graph advantages
+================================================================================
+
+The standard way of launching work on GPUs via streams incurs a small overhead
+for each iteration of the operation involved. For kernels that perform large
+operations during an iteration this overhead usually is negligible. However
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: operations during an iteration this overhead is usually negligible. However

File: docs/understand/hipgraph.rst [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+many workloads, including scientific simulations and AI, a kernel performs a
```
Comment (randyh62) [inline docs/understand/hipgraph.rst]: in many workloads, such as scientific simulations and AI, a kernel performs a

**Review by randyh62 — APPROVED** [review]
Just noticed this is consolidated under PR #3585 

I think this PR can be closed in favor of that one.

File: docs/data/understand/hipgraph/hip_graph.svg
Comment (MKKnorr) [reply docs/data/understand/hipgraph/hip_graph.svg:1]: I agree, that the colour choice was not the best, but I tried to keep it close to the existing colour schemes. I replaced it with a (hopefully) nicer colour set

File: docs/how-to/programming_manual.md (hunk 8 above)
Comment (MKKnorr) [reply docs/how-to/programming_manual.md:147]: I agree, that it's not really needed here, but the programming manual is going to be overhauled and probably incorporated into the runtime api how-to soon, as it is just an amalgamation of snippets on different HIP-topics.

File: docs/understand/hipgraph.rst (hunk 9 above)
Comment (MKKnorr) [reply docs/understand/hipgraph.rst]: Done

File: docs/understand/hipgraph.rst (hunk 10 above)
Comment (MKKnorr) [reply docs/understand/hipgraph.rst]: Fixed

File: docs/understand/hipgraph.rst (hunk 11 above)
Comment (MKKnorr) [reply docs/understand/hipgraph.rst]: Fixed

**Comment by MKKnorr** [comment]
Superseded by #3585

**Pull request closed without merging**